pkg/controller/sparkapplication/controller.go

```diff
@@ -501,20 +507,33 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
 func (c *Controller) updateSparkApplicationStatusWithRetries(
 	original *v1alpha1.SparkApplication,
 	updateFunc func(*v1alpha1.SparkApplicationStatus)) *v1alpha1.SparkApplication {
-	toUpdate := original.DeepCopy()
-
 	var lastUpdateErr error
 	for i := 0; i < maximumUpdateRetries; i++ {
-		updated, err := c.tryUpdateStatus(original, toUpdate, updateFunc)
+		toUpdate := original.DeepCopy()
```

Comment:
Since the client always return a copy, the original approach that makes a copy once outside the loop and assign what the client returns to `toUpdate` is more desirable.

Reply:
As I mentioned, this works fine if we aren't publishing metrics.

For metrics as well as to verify if the stateTransition is valid, we need to maintain the last returned by crdClient. 
We can set `toUpdate` as the last-state returned by the crdClient but then in the next attempt, we also do `	updateFunc(&toUpdate.Status)`  at https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/pull/227/files/f600336b353287b0cd983148071c2bf9640b1135#diff-b7eea120d7ac0e8587781b863b7db8e4R516 . At this point, we lose the last state returned by the `crdClient`. 

For us to  validate `isValidDriverStateTransition` and post metrics, we need to maintain that last state returned by the `crdClient` which is what `original` will be at this point ? 


I might be missing something but this is my understanding of the logic here. I be happy to jump on a quick call if need be.

Reply:
Makes sense.